Ambit/Source/Ambit/Mode/ConfigImportExport.spec.cpp

```diff
@@ -1017,4 +1017,133 @@ void ConfigImportExportSpec::Define()
+
+        AfterEach([this]()
+        {
+            GLevelEditorModeTools().DeactivateMode(FAmbitMode::EM_AmbitModeId);
+            Exporter = nullptr;
+            JsonContent = "";
+        });
+    });
+
+    Describe("OnExportGltf()", [this]()
+    {
+        BeforeEach([this]()
+        {
+            // Generate World
+            World = FAutomationEditorCommonUtils::CreateNewMap();
+
+            // Create Mode Settings
+            GLevelEditorModeTools().ActivateMode(FAmbitMode::EM_AmbitModeId);
+
+            FAmbitMode* AmbitMode = FAmbitMode::GetEditorMode();
+            check(AmbitMode);
+
+            // Generate Exporter
+            Exporter = NewObject<UMockableConfigImportExport>();
+
+            auto MockListBuckets = []() -> TSet<FString>
+            {
+                TSet<FString> setBuckets;
+                setBuckets.Add("BucketName");
+                return setBuckets;
+            };
+            Exporter->SetMockS3ListBuckets(MockListBuckets);
+
+            auto MockCreateBucket = [](const FString& Region, const FString& BucketName) -> void
+            {
+                return;
+            };
+            Exporter->SetMockS3CreateBucket(MockCreateBucket);
+
+            auto MockS3FileUpload = [](const FString& Region, const FString& BucketName, const FString& ObjectName, const FString& FilePath) -> bool
+            {
+                return true;
+            };
+            Exporter->SetMockS3FileUpload(MockS3FileUpload);
+        });
+
+        It("Should fail if no mesh to export", [this]()
+        {
+            AddExpectedError("Cannot find a static mesh to export.", EAutomationExpectedErrorFlags::Exact, 1);
+            Exporter->OnExportGltf();
+        });
+
+        It("Should succeed after test map is loaded", [this]()
+        {
+            // Load the test map.
+            // Loading this map resets some aspects of AmbitMode and the Exporter objects. These need to be reset again
+            // in order for the test to succeed.
+            FAutomationEditorCommonUtils::LoadMap("/Ambit/Test/Maps/ProceduralPlacementTestMap");
+
+            // Create Mode Settings
+            GLevelEditorModeTools().ActivateMode(FAmbitMode::EM_AmbitModeId);
+
+            FAmbitMode* AmbitMode = FAmbitMode::GetEditorMode();
+            check(AmbitMode);
+
+            // This function has been set in BeforeEach() but loading a new map causes it to be unset. Most likely has
+            // to do with state management in UE. Resetting this function should stop any exceptions from being thrown.
+            auto MockS3FileUpload = [](const FString& Region, const FString& BucketName, const FString& ObjectName, const FString& FilePath) -> bool
+            {
+                return true;
+            };
+            Exporter->SetMockS3FileUpload(MockS3FileUpload);
+
+            const FReply Reply = Exporter->OnExportGltf();
+
+            TestTrue("Reply should be true on return", Reply.IsEventHandled());
```

Comment:
Testing the Event being handled by itself is pretty inaccurate. Why did it return true? Did it hit [499](https://github.com/aws-samples/aws-ambit-scenario-designer-ue4/blob/main/Ambit/Source/Ambit/Mode/ConfigImportExport.cpp#L499), [543](https://github.com/aws-samples/aws-ambit-scenario-designer-ue4/blob/main/Ambit/Source/Ambit/Mode/ConfigImportExport.cpp#L543), [563](https://github.com/aws-samples/aws-ambit-scenario-designer-ue4/blob/main/Ambit/Source/Ambit/Mode/ConfigImportExport.cpp#L563), or [592](https://github.com/aws-samples/aws-ambit-scenario-designer-ue4/blob/main/Ambit/Source/Ambit/Mode/ConfigImportExport.cpp#L592)? 

We can use the lambda we are creating at MockS3FileUpload to validate that the function was hit the amount of times we expected, and has the expected results from the parameters. This gives us much more knowledge about that the outputs we are passing are correct and expected.

Reply:
Using `LogErrorAndPopup` in `OnExportGltf()` helps alert UE to any issues that have popped up. In the unit test framework, since there are no errors encountered the conclusion is that the function has reached the end without any issues along the way. Had it hit 499, 543, 563 along the way, the unit test would have failed. This can be seen in the first test which fails when there is no static mesh to be exported (543)

Testing for the other failures would require the MockConfigImportExport to be expanded further.

Reply:
While this is true that we can find it didn't hit the other cases through the logging, we should be very cautious on how fragile this is. 499, for example, does log, but `OnExportGltf()` is not the one who does so, which means their implementation could change and still be ambiguous here. (Unless we have a test case validating that it expects an error with 499, which ensures we have no ambiguity.) This is less of a concern with that point, though. 

That said, the lack of detail behind the middle-parts of the function are still very concerning. We have no mechanism in place to ensure that our outputs of the function are correct through each run. What (if anything), did our function intend to upload to S3 and the details about it? 
It may be better to separate these out to its own test (one for event handled and one for contents). Ideally also one for no OS selected and one where both OS are selected to check between the main cases.

Reply:
I agree with Brandon's comments. I don't know that I'd call this test fragile–it's actually a fine test on its own but only describes one small aspect of the expected behavior of the method. What I expect is more tests that confirm the other expected behaviors. I expect tests for the following...

- Did it call the S3 client API properly?
- Did it write to the correct local directory?
- Does export the correct format (.gltf vs. .glb) based on user input?
- Does it display error popups for each discrete error condition?

You also have some logic in your implementation validating Houdini Asset compatibility and ignoring actors that are pending deletion or not part of the world. Those behaviors need unit test coverage as well.

**I don't think we have to make a part of the scope of this PR, but it's important that the team realize we have some serious gaps in test coverage here.**

I know not all of us are following a test-driven development approach, but even so it's good to keep the TDD maxim of "do the simplest thing that will make the test past" in the back of our minds. Right now, if we replaced the implementation of `OnExportGltf()` with one that just returned a FReply::Handled() response our tests would think everything was fine!